experimental/transliterator_parser/src/parse.rs

```diff
@@ -0,0 +1,1516 @@
+            }
+            // we skipped whitespace and comments, so any other chars must be part of a rule
+            rules.push(self.parse_rule()?);
+        }
+
+        Ok(rules)
+    }
+
+    // expects a rule
+    fn parse_rule(&mut self) -> Result<Rule> {
+        match self.must_peek_char()? {
+            Self::SPECIAL_START => self.parse_filter_or_transform_rule(),
+            // must be a conversion or variable rule
+            _ => self.parse_conversion_or_variable_rule(),
+        }
+    }
+
+    // any rules starting with '::'
+    fn parse_filter_or_transform_rule(&mut self) -> Result<Rule> {
+        // Syntax:
+        // '::' <unicodeset> ';'                  # global filter
+        // '::' '(' <unicodeset> ')' ';'          # global inverse filter
+        // '::' <single-id> (<single-id>)? ';'    # transform rule
+
+        self.consume(Self::SPECIAL_START)?;
+        self.consume(Self::SPECIAL_START)?;
+
+        // because all three options can start with a UnicodeSet, we just try to parse everything
+        // into options, and assemble at the end
+
+        let (forward_filter, forward_basic_id, reverse_filter, reverse_basic_id, has_reverse) =
+            self.parse_filter_or_transform_rule_parts()?;
+
+        self.skip_whitespace();
+
+        // the offset of ';'
+        let meta_err_offset = self.must_peek_index()?;
+        self.consume(Self::RULE_END)?;
+
+        // try to assemble the rule
+        // first try global filters
+        match (
+            forward_filter.is_some(),
+            forward_basic_id.is_some(),
+            reverse_filter.is_some(),
+            reverse_basic_id.is_some(),
+        ) {
+            (true, false, false, false) => {
+                // by match, forward_filter.is_some() is true
+                #[allow(clippy::unwrap_used)]
+                return Ok(Rule::GlobalFilter(forward_filter.unwrap()));
+            }
+            (false, false, true, false) => {
+                // by match, reverse_filter.is_some() is true
+                #[allow(clippy::unwrap_used)]
+                return Ok(Rule::GlobalInverseFilter(reverse_filter.unwrap()));
+            }
+            _ => {}
+        }
+
+        // if this is not a global (inverse) filter rule, this must be a transform rule
+
+        // either forward_basic_id or reverse_basic_id must be nonempty
+        if forward_basic_id.is_none() && reverse_basic_id.is_none() {
+            return Err(PEK::InvalidId.with_offset(meta_err_offset));
+        }
+
+        if !has_reverse {
+            // we must have a forward id due to:
+            //  1. !has_reverse implying reverse_basic_id.is_none()
+            //  2. the above none checks implying forward_basic_id.is_some()
+            // because this is difficult to verify, returning a PEK::Internal anyway
+            // instead of unwrapping, despite technically being unnecessary
+            let forward_basic_id = forward_basic_id.ok_or(PEK::Internal)?;
```

Comment:
thought: add a `&'static str` to `Internal`

Reply:
I like the idea, added it as a polish item to #3736